### ISSUE TITLE
New Annotation type with Spatial labels

### ISSUE THREAD

**Issue by magdalenafuentes** [soundata/soundata#58]
As @ iranroman pointed out in #57, some spatial audio datasets have data in the format:

`time-frame number`, `active class index`, `event number index`, `azimuth`, and `elevation` 

* `time-frame` is a timestamp
* `active class index` is the class ID of the event
* `event number index` is a sequence of increasing integers to indicate the order of the events and avoid confusion between overlapping events of the same class.
* `azimuth` and `elevation` are self-explanatory (in units of degrees). 

This is true for:

* TAU NIGENS SSE 2021

* TAU NIGENS SSE 2020

* TAU SSE 2019. Annotations where are `frame number`, `active class index`, `event number index`, `azimuth`, and `elevation`, `distance`.

@iranroman suggested to add a class `SpatialEvents`, that I imagine will be an extended version of `Events`, something like:

* intervals (from one timestamp to the next)
* labels (class index as str)
* confidence (1.0 if it's manual, same as events)
* position (azimuth, elevation, distance), where is mandatory to have one of them but not all?

@iranroman I think that to get the intervals down you'd have to look at event number index and keep track of when a class appears and goes away, does it make sense?

@justinsalamon @pzinemanas @mrocamora thoughts? @pzinemanas this is also relevant to `DCASE_models` since these are associated to DCASE Task 3 2021 (and same task in previous years)

**Comment by magdalenafuentes** [comment]
Moving the conversation here, adding @iranroman 's comment here:

"I think what we need is a `SpatialEvents` class that accounts for `azimuth`, `elevation`, and `distance` of events. This would be a simple `Events` class spinoff. 

```
class SpatialEvents(Annotation):
    """Events class

    Attributes:
        intervals (np.ndarray): (n x 2) array of intervals
            (as floats) in seconds in the form [start_time, end_time]
            with positive time stamps and end_time >= start_time.
        labels (list): list of event labels (as strings)
        azimuth (list): list of azimuth degree (as int in the rante [0 360])
        elevation (list): list of elevation degree (as int in the range [0, 180])
        distance (list): list of distance in meters (as float, or None)
        confidence (np.ndarray or None): array of float confidence values
            in the range [0, 1]
    """

    def __init__(self, intervals, labels, azimuth, elevation, distance, confidence=None) -> None:
        validate_array_like(intervals, np.ndarray, float)
        validate_array_like(labels, list, str)
        validate_array_like(azimuth, list, int)
        validate_array_like(elevation, list, int)
        validate_array_like(distance, list, float, none_allowed=True)
        validate_array_like(confidence, np.ndarray, float, none_allowed=True)
        validate_lengths_equal([intervals, labels, azimuth, elevation, distance, confidence])
        validate_intervals(intervals)
        validate_confidence(confidence)
        validate_azimuth(azimuth)
        validate_elevation(elevation)

        self.intervals = intervals
        self.labels = labels
        self.confidence = confidence
        self.confidence = azimuth
        self.elevation = elevation
        self.distance = distance
```

For the specific case of TAU NIGENS 2020 and 2021, those dataset modules can define their own `IndexedSpatialEvents` class that inherits from the `SpatialEvents` class and accounts for the `event number index`, which I'm not sure will be broadly relevant for more datasets. "

**Comment by magdalenafuentes** [comment]
@iranroman I like your proposal, seems general enough. I imagine that for the class to be general enough maybe we want to have `azimuth` and `elevation` as optional as well, and then enforce that the class has at least one of `azimuth`, `distance` or `elevation`. What do you think?

**Comment by magdalenafuentes** [comment]
What do they use in Task 3 to evaluate the models do you know? @iranroman? I think that ideally we want `soundata` to load things in memory in a way that they are compatible with e.g. `sed_eval` directly

**Comment by iranroman** [comment]
@magdalenafuentes it is true that out of the three position (azimuth, elevation, distance) only one is required. Some datasets only explore one or two of these dimensions for localization. 

 _think that to get the intervals down you'd have to look at event number index and keep track of when a class appears and goes away, does it make sense?_

This is also true. But for the DCASE datasets I will have to parse their timestamps. I do believe it's worth it though. If we do this, then we do not need the `event number index` at all. 

For DCASE, these are their metrics [https://github.com/sharathadavanne/seld-dcase2021/blob/master/metrics/SELD_evaluation_metrics.py](https://github.com/sharathadavanne/seld-dcase2021/blob/master/metrics/SELD_evaluation_metrics.py) but I have not played around with those myself yet. They include F-Score and Error Rate for event detection, as well as localization error and localization recall. [http://dcase.community/challenge2021/task-sound-event-localization-and-detection#evaluation](http://dcase.community/challenge2021/task-sound-event-localization-and-detection#evaluation).

**Comment by iranroman** [comment]
I studied the `seld-dcase2021` code to understand how they process the annotations. They convert the csv values into a dictionary where the `time-frames` are the keys, and the values are a list with the class category, coordinates and `event number index`.

https://github.com/sharathadavanne/seld-dcase2021/blob/b279508ff4752f25cd006d503d78da519c86764e/cls_feature_class.py#L270-L291

Then they use the dictionary values to create four matrices `Ntime X NClasses`, one with one-hot encodings for each class, and three more with the cartesian X, Y, and Z coordinates each for each class. 

 https://github.com/sharathadavanne/seld-dcase2021/blob/b279508ff4752f25cd006d503d78da519c86764e/cls_feature_class.py#L147-L171

In short: I think our soundata loader will be more general than the way they load their data, which is a good thing. Is there a specific set of metrics/functions we want `SpatialEvents` to work with?

**Comment by magdalenafuentes** [comment]
Hey @iranroman thanks a lot for taking a look at that! The discussion about `SpatialEvents` is making me think of adding a few extensions to the annotation types that were introduced recently in `mirdata` and make sense here as well:

1. _**Annotation units**_: If we want to be "general enough" with annotations such as `azimuth`, `elevation` or `distance`, I'm starting to feel that we should include the unit as part of the annotation. So if in the future we want to include a dataset that has some sort of `azimuth` information in other than degrees in `0, ..., 360` (e.g. radians) we can reuse the same annotation type. Maybe nobody uses radians but I think `units` won't hurt, and on the contrary they'll clarify!
2. _**Annotation format converters**:_ It seems to me that the formats you're mentioning are useful for the `DCASE` challenge to train the model, but maybe people would want the annotations in other formats for e.g. evaluation or analysis. A nice way of doing this is to add functions `to_class_matrix`, `to_cartesian_matrix` to the annotation class, and let the user choose what they need. 

Do you think this would include the use cases you're seeing with these datasets?

**Comment by iranroman** [comment]
I agree! These two points would be very useful for microphone array datasets in general. We can include them. I'll be happy to write them. What does @justinsalamon (and others) think?

**Comment by justinsalamon** [comment]
Hey, just catching up now.

My first question is whether it makes sense to create a new annotation type over expanding the existing `Event` type. For example, the spatial fields could be added to the event type as optional fields, such that it works as it does now for non-spatial events but can also support spatial events.

What are the arguments in favor of creating a new type of expanding the existing one? I ask because I want to make sure we:
* avoid "annotation sprawl" where small differences in datasets lead to the creation of more and more annotation types
* keep the API as simple as possible
* minimize technical debt

@iranroman @magdalenafuentes Let me know your thoughts, thanks!

**Comment by iranroman** [comment]
I agree that expanding the existing `Event` type should be good enough. Is there a way for us to make the Event type flexible enough to have it include ANY information that comes with the dataset? Right now we are talking about adding the spatial coordinate information to the event, but maybe there'll be the need for more flexibility down the road as we add more datasets. 

Alternatively, maybe `Events` is already is flexible enough if I use it as the parent of my own class. I could just define my own class in the loader and add the needed attributes (@magdalenafuentes originally suggested that I do that). 

But something tells me that the spatial information is important enough to be built-in the main `Events` or have its own `SpatialEvents` class, though. 

It's a matter of perspective. Just keep in mind that I will be adding loaders for many microphone array datasets, many of which include the spatial information of events.

**Comment by justinsalamon** [comment]
Thanks @iranroman. I like @magdalenafuentes 's idea of using `Events` as a parent class - this way you can add the functionality you need without having to reimplement/duplicate all the functionality that Event already provides. This minimizes technical debt (no need for duplicate tests or maintenance) and helps us keep the API clean (everything that works for an `Event` will work for your inherited class too).

I would favor keeping "top level" types unique in terms of the type of data they represent, e.g., Tag (describes the whole recording) vs Event (describes a moment in time in the recording). A spatial event is a type of event, rather than a fundamentally different type of annotation. So inheritance feel like the right thing to do here: it takes advantage of the commonalities between Event and SpatialEvent while allowing the flexibility for SpatialEvent to add the fields that it needs.

@magdalenafuentes @iranroman thoughts?

Thanks!

**Comment by iranroman** [comment]
I agree. Just to confirm, you are saying that each "spatial" dataset loader will independently define this `SpatialEvents` class inheriting from `Events`, and will adjust it according to what its needs, right?

**Comment by magdalenafuentes** [comment]
Hi both, I'm in sync with the suggestion of a child class `SpatialEvent` that inherits from `Event`. However, I think we should implement it in `annotations.py` because it would be used in many datasets, not only the ones that @iranroman is planning to add but also @nkundiushuti might add some others in the future. 

Also I think it will simplify code, i.e. avoid to have that class implementation multiple times in the library in the different loaders. I have a clear idea in mind so I can sketch a PR tomorrow and see if we all agree. Thoughts? @justinsalamon

**Comment by justinsalamon** [comment]
sgtm @magdalenafuentes 👍

**Comment by iranroman** [comment]
Given everything we have discussed, I believe that `SoundEvents` should have the following arguments:

```
    Attributes:
        intervals (np.ndarray): (n x 2) array of intervals
            (as floats) in seconds in the form [start_time, end_time]
            with positive time stamps and end_time >= start_time.
        labels (list): list of event labels (as strings)
        time_step (float): the time-step between events over time
        position (list): list with arrays, or None. 
            None indicates no position information.
            a 1x2 array indicates the azimuth and elevation of a spatially static event.
            a Tx2 array indicates the azimuth and elevation of a moving event, where T is
                the number of times_steps the event is present in the scene. None indicates
                that the event is temporarily untraceable. T == 1+ (end_time - start_time)/time_step
            azimuth values must have a range less than 360 (degrees) or 2*pi (radians)
            elevation values mut have a range less than 180 (degrees) or pi (radians)
        active_time_steps (list): list with arrays, True, or None.
             None indicates no activity information at the level of time_steps
             True indicates that the event is always active
             a T-long 1D array with True or False values, indicating whether the event is active. 
        confidence (np.ndarray or None): array of confidence values, float in [0, 1]
        labels_unit (str): labels unit, one of LABELS_UNITS
        position_unit (str): labels unit, one of POSITION_UNITS
        intervals_unit (str): intervals unit, one of TIME_UNITS
```

Let's please come to an agreement soon. I want to add three DCASE dataset with this structure as soon as possible to start working with these loaders myself.

**Comment by iranroman** [comment]
`position` will need to de 3D, actually, because position information may also include distance (or radius). Therefore:
```
        position (list): list with arrays, or None. 
            None indicates no position information.
            a 1x3 array indicates the azimuth, elevation and distance of a spatially static event.
            a Tx3 array indicates the azimuth, elevation and distance of a moving event, where T is
                the number of times_steps the event is present in the scene. None indicates
                that the event is temporarily untraceable. T == 1+ (end_time - start_time)/time_step
            azimuth values must have a range less than 360 (degrees) or 2*pi (radians)
            elevation values must have a range less than 180 (degrees) or pi (radians)
            distance values must be positive.  
```
What do you think @magdalenafuentes and @justinsalamon? Please also see my previous comment.

**Comment by iranroman** [comment]
Reviving this conversation as I am working on loaders with `SpatialEvents`. I have one question: is there a way to expand the `JAMS` format to account for spatial sound events? I have heard from DCASE community members things along the lines of:

> can we use a standardized output format for sound event location and detection labels? For example is the JAMS format or any other similar JSON-like format appropriate for having a standardized annotation format?

Please let me know your thoughts. Thanks!

**Comment by magdalenafuentes** [comment]
Hi @iranroman

Thanks for reviving this and bringing up the standardization issue with spatial annotations. I think we could take this as an attempt to standardize this annotations, and draft the structure and consistency checks to then move that to `JAMS`? ccing @justinsalamon @bmcfee to share their thoughts, and maybe we should involve other people from the DCASE community?

**Comment by bmcfee** [comment]
Sigh -- yeah, we really need to get on with the upgrades to JAMS.  I just haven't had the time or capacity to run that show lately.  We do have some fairly concrete plans that would make this sort of thing possible though.
